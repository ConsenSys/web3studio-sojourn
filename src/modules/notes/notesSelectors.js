import { createSelector } from 'reselect';
import moment from 'moment';
import uuid from 'uuid/v4';

/**
 * Select the top level meat grinder state
 *
 * @param {Object} state - Redux state
 * @returns {Object} - meat grinder state
 */
const selectNotes = state => state.notes.notes;

/**
 * Select the current map of vault records
 *
 * @param {Object} state - Redux state
 * @returns {Object} - vault record map
 */
const selectVaultRecords = state => state.notes.vaultRecords;

/**
 * Select the current vault record identifier
 *
 * @param {Object} state - Redux state
 * @param {Object} props - React element props
 * @returns {string} - vault record hash
 */
const selectCurrentVaultRecordHash = (state, props) => props.vaultRecord;

/**
 * Select the current vault record
 *
 * @param {Object} state - Redux state
 * @param {Object} props - React element props
 * @returns {Object} - A vault record
 */
export const selectCurrentVaultRecord = createSelector(
  selectVaultRecords,
  selectCurrentVaultRecordHash,
  (vaultRecords, vaultRecordHash) => {
    return vaultRecords[vaultRecordHash];
  }
);

/**
 * Select the noteId passed by navigation param
 *
 * @param {Object} state - Redux state object
 * @param {Object} props - props passed to component
 * @returns {string} - noteId
 */
const selectCurrentNoteId = (state, props) =>
  props.navigation.getParam('noteId') || uuid();

/**
 * Select the current note passed by navigation param `noteId`
 *
 * @param {Object} state - Redux state object
 * @param {Object} props - props passed to component
 * @returns {Object} - a note object
 */
export const selectCurrentNote = createSelector(
  selectNotes,
  selectCurrentNoteId,
  (notes, noteId) => notes[noteId] || { id: noteId, title: null, body: null }
);

/**
 * Select a sorted list of notes with computed properties for display to the
 * user.
 *
 * @param {Object} state - Redux state object
 * @returns {Array<Object>} - An array of note objects
 */
export const selectNotesDisplayList = createSelector(selectNotes, notes =>
  Object.keys(notes)
    .map(id => notes[id])
    .map(note => ({
      ...note,
      body: note.body && note.body.trim(),
      calendarLastModified: moment(note.lastModified).fromNow(true)
    }))
    .sort((noteA, noteB) => noteB.lastModified - noteA.lastModified)
);
