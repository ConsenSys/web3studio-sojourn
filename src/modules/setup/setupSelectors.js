import { createSelector } from 'reselect';

/**
 * Selects sign in state
 *
 * @param {Object} state - redux state
 * @returns {Object} Sign in redux state
 */
const selectSetup = state => state.setup;
const selectUPort = createSelector(
  selectSetup,
  setup => setup.uport
);

export const selectAddress = createSelector(
  selectUPort,
  uport => uport.address
);

export const hasSetupUport = createSelector(
  selectAddress,
  address => !!address
);
