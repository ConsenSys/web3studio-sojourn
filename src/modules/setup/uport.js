import configureUportConnect from 'react-native-uport-connect';
import {
  UPORT_APP_NAME,
  UPORT_APP_ADDRESS,
  UPORT_PRIVATE_KEY
} from 'react-native-dotenv';
import DeviceInfo from 'react-native-device-info';
import Web3 from 'web3';

const { uport, MNID } = configureUportConnect({
  appName: UPORT_APP_NAME,
  appAddress: UPORT_APP_ADDRESS,
  privateKey: UPORT_PRIVATE_KEY
});

const isEmulator = DeviceInfo.isEmulator();

// uPort wont run on an emulator, replacing interesting
// functionality with ganache
const web3Provider = !isEmulator
  ? uport.getProvider()
  : new Web3.providers.WebsocketProvider('http://127.0.0.1:8545');

const web3 = new Web3(web3Provider);

const networkId = !isEmulator
  ? web3.utils.hexToNumberString(uport.network.id)
  : '42';

const requestCredentials = !isEmulator
  ? async () => {
      const payload = await uport.requestCredentials();

      return {
        address: MNID.decode(payload.networkAddress).address
      };
    }
  : async () => {
      const accounts = await web3.eth.getAccounts();

      return {
        address: accounts[0]
      };
    };

export { web3, requestCredentials, networkId };
