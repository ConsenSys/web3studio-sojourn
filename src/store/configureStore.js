import { createStore, applyMiddleware } from 'redux';
import { persistStore, persistReducer } from 'redux-persist';
import thunk from 'redux-thunk';
import secureStorage from './secureStorage';
import stateReconciler from 'redux-persist/lib/stateReconciler/autoMergeLevel2';

import rootReducer from './reducer';

const persistConfig = {
  key: 'StoreRoot',
  storage: secureStorage,
  stateReconciler
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export default () => {
  const store = createStore(persistedReducer, applyMiddleware(thunk));

  const persistor = persistStore(store);

  if (module.hot) {
    module.hot.accept(() => {
      // This fetch the new state of the above reducers.
      const nextRootReducer = require('./reducer');

      store.replaceReducer(persistReducer(persistConfig, nextRootReducer));
    });
  }

  return { store, persistor };
};
