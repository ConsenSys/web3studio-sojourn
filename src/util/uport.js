import configureUportConnect from 'react-native-uport-connect';
import {
  UPORT_APP_NAME,
  UPORT_APP_ADDRESS,
  UPORT_PRIVATE_KEY
} from 'react-native-dotenv';

export const { uport, MNID } = configureUportConnect({
  appName: UPORT_APP_NAME,
  appAddress: UPORT_APP_ADDRESS,
  privateKey: UPORT_PRIVATE_KEY
});
