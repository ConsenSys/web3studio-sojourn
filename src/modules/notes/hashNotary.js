import { web3, networkId } from '../signIn/uport';
import DeviceInfo from 'react-native-device-info';

let HashNotaryContract = require('../../contracts/HashNotary.json');

if (DeviceInfo.isEmulator()) {
  HashNotaryContract = require('../../../build/contracts/HashNotary.json');
}

const currentContractNetwork = HashNotaryContract.networks[networkId];

/* istanbul ignore next */
if (!currentContractNetwork && __DEV__) {
  // Logging error for mis-configured dev environment
  // eslint-disable-next-line no-console
  console.error(
    'Unable to find deployed HashNotary contract on network ',
    networkId
  );
}

export default new web3.eth.Contract(
  HashNotaryContract.abi,
  currentContractNetwork.address
);
