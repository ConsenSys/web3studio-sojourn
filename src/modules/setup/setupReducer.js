import { requestCredentials } from './uport';

export const SETUP_UPORT_SUCCESSFUL = 'sojourn/sign-in/SETUP_UPORT_SUCCESSFUL';
export const SETUP_UPORT_FAILURE = 'sojourn/sign-in/SETUP_UPORT_FAILURE';

/**
 * Setup uPort success action generator
 *
 * @param {Object} payload - uPort Payload
 * @returns {Object} - redux action
 */
export const uportSetupSuccess = payload => ({
  type: SETUP_UPORT_SUCCESSFUL,
  payload
});

/**
 * Setup uPort failure action generator
 *
 * @param {Error} error - sign in error
 * @returns {Object} - redux action
 */
export const uportSetupFailure = error => ({
  type: SETUP_UPORT_FAILURE,
  error
});

/**
 * Sign in thunk. Uses uPort to Authenticate the user
 *
 * @returns {Function} - redux thunk
 */
export const setupUport = () => dispatch =>
  requestCredentials()
    .then(payload => dispatch(uportSetupSuccess(payload)))
    .catch(error => dispatch(uportSetupFailure(error)));

export const INITIAL_STATE = {
  uport: {}
};

/**
 * Sign in redux reducer
 *
 * @param {Object} state - Current redux state
 * @param {Object} action - redux action
 * @returns {Object} - new redux state
 */
export const reducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case SETUP_UPORT_SUCCESSFUL:
      return {
        ...state,
        uport: {
          ...action.payload
        }
      };

    case SETUP_UPORT_FAILURE:
      return {
        ...state,
        uport: INITIAL_STATE.uport,
        error: action.error
      };
  }
  return state;
};
