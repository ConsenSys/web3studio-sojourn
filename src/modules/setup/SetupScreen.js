import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { StyleSheet } from 'react-native';
import { Button, H2, Text, Container, Content } from 'native-base';
import { connect } from 'react-redux';
import { setupUport } from './setupReducer';
import { hasSetupUport } from './setupSelectors';

/**
 * Sign in screen. Allows users to log in with uPort.
 */
export class SetupScreen extends Component {
  static propTypes = {
    navigation: PropTypes.object,
    hasSetupUport: PropTypes.bool
  };

  /**
   * Called on prop update
   */
  componentDidUpdate() {
    this.navigateOnSignIn();
  }

  /**
   * Called on mount
   */
  componentDidMount() {
    this.navigateOnSignIn();
  }

  /**
   * Navigates to Application router when signed in
   */
  navigateOnSignIn = () => {
    const { navigation, hasSetupUport } = this.props;

    if (hasSetupUport) {
      navigation.navigate('App');
    }
  };

  /**
   * Dispatches sign in redux action
   */
  uPortLogin = async () => {
    const { setupUport } = this.props;

    setupUport();
  };

  /**
   * Renders the component.
   *
   * @returns {React.Element} Element to render.
   */
  render() {
    return (
      <Container testID="setup-screen">
        <Content contentContainerStyle={styles.content}>
          <H2 style={styles.welcome}>Sojourn</H2>
          <Text style={styles.welcome}>
            Sojourn uses uPort to make your notes tamper-resistant. Add or
            create your uPort Identity.
          </Text>
          <Button style={styles.button} onPress={this.uPortLogin}>
            <Text>Continue with uPort</Text>
          </Button>
        </Content>
      </Container>
    );
  }
}

const styles = StyleSheet.create({
  content: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#fff'
  },
  welcome: {
    textAlign: 'center',
    marginTop: 20
  },
  button: {
    textAlign: 'center',
    alignSelf: 'center',
    marginTop: 40
  }
});

/**
 * Redux connect mappings
 *
 * @param {Object} state - redux state
 * @returns {{isSignedIn}} - injected props
 */
const mapStateToProps = state => ({
  hasSetupUport: hasSetupUport(state)
});

const mapDispatchToProps = {
  setupUport
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SetupScreen);
