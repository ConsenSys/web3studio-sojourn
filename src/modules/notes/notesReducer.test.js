import { EventEmitter } from 'events';
import { promisify } from 'util';
import {
  notarizeAndSaveNoteToVault,
  notaryStatus,
  noteChanged
} from './notesReducer';
import hashNotary from './hashNotary';
import configureStore from '../../store/configureStore';
import { web3 } from '../setup/uport';
import { uportSetupSuccess } from '../setup/setupReducer';
import { selectCurrentVaultRecord, selectCurrentNote } from './notesSelectors';

jest.mock('./hashNotary');

/**
 * Creates a mocked contract method
 *
 * @returns {Function} - A Mocked web3 Contract method
 */
const mockContractMethod = () => {
  const mockSendEmitter = new EventEmitter();
  const mockCall = jest.fn();

  const mockMethod = jest.fn().mockReturnValue({
    send: jest.fn().mockReturnValue(mockSendEmitter),
    call: mockCall
  });

  mockMethod._mockSendEmitter = mockSendEmitter;
  mockMethod._mockCall = mockCall;

  return mockMethod;
};

describe('Notes reducer', () => {
  const txHash = '0x0';
  const note = { id: '10', title: '10 is best', body: 'We all love David' };
  const noteString = JSON.stringify(note);
  const noteHash = web3.utils.sha3(noteString);
  const expectedTimestamp = Math.floor(Date.now() / 1000);
  let store;

  beforeEach(() => {
    store = configureStore().store;

    // Mock a log in with uport
    store.dispatch(
      uportSetupSuccess({
        networkAddress: '2nQtiQG6Cgm1GYTBaaKAgr76uY7iSexUkqX'
      })
    );

    hashNotary.methods.save = mockContractMethod();
    hashNotary.methods.getTimestamp = mockContractMethod();

    hashNotary.methods.getTimestamp._mockCall.mockReturnValue(
      expectedTimestamp
    );

    fetch.mockResponse(JSON.stringify({ Key: 'QM1', Size: 99 }));
  });

  it('saves a status and a txHash for later on notary start', async () => {
    store.dispatch(notarizeAndSaveNoteToVault(note));

    hashNotary.methods.save._mockSendEmitter.emit('transactionHash', txHash);

    // Skip a tick for callbacks to pop off
    await promisify(setImmediate)();
    const vaultRecord = selectCurrentVaultRecord(store.getState(), {
      vaultRecord: noteHash
    });

    expect(vaultRecord).toEqual(
      expect.objectContaining({
        noteHash,
        notary: {
          status: notaryStatus.pending,
          txHash
        }
      })
    );
  });

  it('saves an error and status on failure', async () => {
    const error = new Error('A wild Dalek appeared');

    store.dispatch(notarizeAndSaveNoteToVault(note));

    hashNotary.methods.save._mockSendEmitter.emit('error', error);

    // Skip a tick for callbacks to pop off
    await promisify(setImmediate)();
    const vaultRecord = selectCurrentVaultRecord(store.getState(), {
      vaultRecord: noteHash
    });

    expect(vaultRecord).toEqual(
      expect.objectContaining({
        noteHash,
        notary: {
          status: notaryStatus.error,
          error
        }
      })
    );
  });

  // If this test case fails, it's likely due to a timeout on a slow machine
  // Check that first to see if it needs increasing
  it('notarizes after 12 confirmations and uploads to IPFS', async () => {
    store.dispatch(notarizeAndSaveNoteToVault(note));

    hashNotary.methods.save._mockSendEmitter.emit('transactionHash', txHash);

    // Simulate 24 confirmations
    new Array(24).fill('').forEach((val, index) => {
      hashNotary.methods.save._mockSendEmitter.emit('confirmation', index + 1);
    });

    // Mocks take about 61 ms, this should cover timeouts
    await promisify(setTimeout)(300);

    const vaultRecord = selectCurrentVaultRecord(store.getState(), {
      vaultRecord: noteHash
    });

    const fileContents = fetch.mock.calls.map(([url, request]) => {
      expect(url).toEqual('https://ipfs.infura.io:5001/api/v0/block/put');

      return request.body;
    });

    expect(fileContents).toMatchSnapshot();
    expect(vaultRecord).toEqual(
      expect.objectContaining({
        noteId: note.id,
        noteHash,
        notary: {
          status: notaryStatus.confirmed,
          txHash,
          timestamp: expectedTimestamp
        },
        ipfs: {
          hashes: new Array(12).fill('QM1'),
          iv: 'AQEBAQEBAQEBAQEBAQEBAQ=='
        }
      })
    );
  });

  it('can update an existing note', () => {
    const params = {
      navigation: {
        getParam: () => note.id
      }
    };

    const updatedNote = {
      ...note,
      title: note.title + ' ' + note.title
    };

    store.dispatch(noteChanged(note));
    expect(selectCurrentNote(store.getState(), params)).toEqual(
      expect.objectContaining(note)
    );

    store.dispatch(noteChanged(updatedNote));
    expect(selectCurrentNote(store.getState(), params)).toEqual(
      expect.objectContaining(updatedNote)
    );
  });
});
