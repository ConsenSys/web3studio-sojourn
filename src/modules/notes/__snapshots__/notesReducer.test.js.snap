// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notes reducer notarizes after 12 confirmations and uploads to IPFS 1`] = `
Array [
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "01100000000017b226964223a223130222c227469746c65223a2231302069732062657374222c22626f6479223a22576520616c6c206c6f7665204461766964227d",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "0217e7e7e7e7f055c171a5c445c4f4e5c525c0a170a121b5c445c4f4e5e170d5e1c1b0d0a5c525c1c111a075c445c291b5e1f12125e1211081b5e3a1f08171a5c03",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "0317e7e7e7e7f055c171a5c445c4f4e5c525c0a170a121b5c445c4f4e5e170d5e1c1b0d0a5c525c1c111a075c445c291b5e1f12125e1211081b5e3a1f08171a5c03",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "041f0f0f0f0f18bd29994d2cad2c1c0d2dcd28499849c95d2cad2c1c0d09983d092958384d2dcd2929f9489d2cad2a795d0919c9cd09c9f8695d0b491869994d28d",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "051f0f0f0f0f18bd29994d2cad2c1c0d2dcd28499849c95d2cad2c1c0d09983d092958384d2dcd2929f9489d2cad2a795d0919c9cd09c9f8695d0b491869994d28d",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "061adadadadacd68fc4c98f978f9c9d8f818fd9c4d9c1c88f978f9c9d8dc4de8dcfc8ded98f818fcfc2c9d48f978ffac88dccc1c18dc1c2dbc88de9ccdbc4c98fd0",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "071adadadadacd68fc4c98f978f9c9d8f818fd9c4d9c1c88f978f9c9d8dc4de8dcfc8ded98f818fcfc2c9d48f978ffac88dccc1c18dc1c2dbc88de9ccdbc4c98fd0",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "081b4b4b4b4b5cf96ddd0968e968584969896c0ddc0d8d1968e96858494ddc794d6d1c7c0969896d6dbd0cd968e96e3d194d5d8d894d8dbc2d194f0d5c2ddd096c9",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "091b4b4b4b4b5cf96ddd0968e968584969896c0ddc0d8d1968e96858494ddc794d6d1c7c0969896d6dbd0cd968e96e3d194d5d8d894d8dbc2d194f0d5c2ddd096c9",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "0a10a0a0a0a0b7128636e2830283b3a2826287e637e666f2830283b3a2a63792a686f797e28262868656e732830285d6f2a6b66662a66657c6f2a4e6b7c636e2877",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "0b10a0a0a0a0b7128636e2830283b3a2826287e637e666f2830283b3a2a63792a686f797e28262868656e732830285d6f2a6b66662a66657c6f2a4e6b7c636e2877",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "0c17d7d7d7d7c065f14195f475f4c4d5f515f09140911185f475f4c4d5d140e5d1f180e095f515f1f1219045f475f2a185d1c11115d11120b185d391c0b14195f00",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "01100000000017b226964223a223130222c227469746c65223a2231302069732062657374222c22626f6479223a22576520616c6c206c6f7665204461766964227d",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "0217e7e7e7e7f055c171a5c445c4f4e5c525c0a170a121b5c445c4f4e5e170d5e1c1b0d0a5c525c1c111a075c445c291b5e1f12125e1211081b5e3a1f08171a5c03",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "0317e7e7e7e7f055c171a5c445c4f4e5c525c0a170a121b5c445c4f4e5e170d5e1c1b0d0a5c525c1c111a075c445c291b5e1f12125e1211081b5e3a1f08171a5c03",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "041f0f0f0f0f18bd29994d2cad2c1c0d2dcd28499849c95d2cad2c1c0d09983d092958384d2dcd2929f9489d2cad2a795d0919c9cd09c9f8695d0b491869994d28d",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "051f0f0f0f0f18bd29994d2cad2c1c0d2dcd28499849c95d2cad2c1c0d09983d092958384d2dcd2929f9489d2cad2a795d0919c9cd09c9f8695d0b491869994d28d",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "061adadadadacd68fc4c98f978f9c9d8f818fd9c4d9c1c88f978f9c9d8dc4de8dcfc8ded98f818fcfc2c9d48f978ffac88dccc1c18dc1c2dbc88de9ccdbc4c98fd0",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "071adadadadacd68fc4c98f978f9c9d8f818fd9c4d9c1c88f978f9c9d8dc4de8dcfc8ded98f818fcfc2c9d48f978ffac88dccc1c18dc1c2dbc88de9ccdbc4c98fd0",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "081b4b4b4b4b5cf96ddd0968e968584969896c0ddc0d8d1968e96858494ddc794d6d1c7c0969896d6dbd0cd968e96e3d194d5d8d894d8dbc2d194f0d5c2ddd096c9",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "091b4b4b4b4b5cf96ddd0968e968584969896c0ddc0d8d1968e96858494ddc794d6d1c7c0969896d6dbd0cd968e96e3d194d5d8d894d8dbc2d194f0d5c2ddd096c9",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "0a10a0a0a0a0b7128636e2830283b3a2826287e637e666f2830283b3a2a63792a686f797e28262868656e732830285d6f2a6b66662a66657c6f2a4e6b7c636e2877",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "0b10a0a0a0a0b7128636e2830283b3a2826287e637e666f2830283b3a2a63792a686f797e28262868656e732830285d6f2a6b66662a66657c6f2a4e6b7c636e2877",
      ],
    ],
  },
  FormData {
    "_parts": Array [
      Array [
        "file1",
        "0c17d7d7d7d7c065f14195f475f4c4d5f515f09140911185f475f4c4d5d140e5d1f180e095f515f1f1219045f475f2a185d1c11115d11120b185d391c0b14195f00",
      ],
    ],
  },
]
`;
